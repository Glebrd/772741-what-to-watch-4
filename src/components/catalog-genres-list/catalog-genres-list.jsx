import React from "react";
import {ActionCreator} from "../../reducer";
import {connect} from "react-redux";
import PropTypes from "prop-types";

export const getGenres = (state) => {
  const genresSet = new Set([`All genres`]);
  state.movies.forEach((movie) => genresSet.add(movie.genre));
  return genresSet;
};

const CatalogGenresList = (props)=>{
  const {currentGenre, onChange, genres} = props;
  return (
    <ul className="catalog__genres-list">
      {[...genres].map((genre) => {
        return (
          <li
            key={genre}
            className= {`catalog__genres-item ${currentGenre === genre ? `catalog__genres-item--active` : ``}`}
          >
            <a href="#" className="catalog__genres-link"
              onClick={(evt) => {
                evt.preventDefault();
                onChange(genre);
              }}

            >{genre}</a>
          </li>
        );
      })}
    </ul>
  );
};

const mapStateToProps = (state) => ({
  currentGenre: state.currentGenre,
  genres: getGenres(state),
});

const mapDispatchToProps = (dispatch) => ({
  onChange(genre) {
    dispatch(ActionCreator.setCurrentGenre(genre));
  },
});

CatalogGenresList.propTypes = {
  currentGenre: PropTypes.string,
  genres: PropTypes.object,
  onChange: PropTypes.func,
};

export {CatalogGenresList};
export default connect(mapStateToProps, mapDispatchToProps)(CatalogGenresList);
