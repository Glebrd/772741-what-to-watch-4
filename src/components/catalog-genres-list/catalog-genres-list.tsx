import * as React from "react";
import withTabs from "../../hocs/with-tabs/with-tabs";

const DEFAULT_ACTIVE_ITEM = `All genres`;

interface Props {
  currentGenre: string;
  genres: Set<string>;
  onChangeTab: (genre: string) => {void};
  onGenreChange: (genre: string) => {void};
  activeTab: string;
}

const CatalogGenresList: React.FunctionComponent<Props> = (props: Props) => {
  const {onGenreChange, onChangeTab, activeTab, genres} = props;
  const currentlyActiveTab = activeTab ? activeTab : DEFAULT_ACTIVE_ITEM;
  return (
    <ul className="catalog__genres-list">
      {[...genres].map((genre) => {
        const currentClassName = `catalog__genres-item ${currentlyActiveTab === genre ? `catalog__genres-item--active` : ``}`;
        return (
          <li
            key={genre}
            className={currentClassName}
          >
            <a href="#" className="catalog__genres-link"
              onClick={(evt) => {
                evt.preventDefault();
                onChangeTab(genre);
                onGenreChange(genre);
              }}

            >{genre}</a>
          </li>
        );
      })}
    </ul>
  );
};

export {CatalogGenresList};
export default withTabs(CatalogGenresList);
