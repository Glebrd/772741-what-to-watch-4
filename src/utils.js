import {ISOStringProperty, MILLISECONDS_IN_SECOND, MINUTES_IN_HOUR} from "./const";

const DELETE_COUNT = 1;

const TimeAbbreviation = {
  HOURS: `h`,
  MINUTES: `m`,
};

export const extend = (firstSource, secondSource) => {
  return Object.assign({}, firstSource, secondSource);
};

export const formatTime = (seconds) => {
  return new Date(seconds * MILLISECONDS_IN_SECOND).toISOString()
    .substr(ISOStringProperty.START_OF_HOURS, ISOStringProperty.LENGTH_OF_HH_MM_SS_SUBSTRING);
};

export const formatDuration = (totalMinutes) =>{
  let hours = Math.floor(totalMinutes / MINUTES_IN_HOUR);
  let minutes = totalMinutes % MINUTES_IN_HOUR;
  return `${hours + TimeAbbreviation.HOURS} ${minutes + TimeAbbreviation.MINUTES}`;
};

export const replaceMovie = (movie, movies) => {
  const index = movies.findIndex((oldMovie) => oldMovie.id === movie.id);
  movies.splice(index, DELETE_COUNT, movie);
  return movies;
};

export const replacePromo = (updatedMovie, promoMovie) => {
  return updatedMovie.id === promoMovie.id ? updatedMovie : promoMovie;
};


