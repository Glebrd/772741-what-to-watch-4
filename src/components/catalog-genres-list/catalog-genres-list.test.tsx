import * as React from "react";
import * as renderer from "react-test-renderer";
import {CatalogGenresList} from "./catalog-genres-list";
import {noOperation} from "../../utils";

const currentGenre = `All genres`;
const genres = new Set([`Drams`, `Sci-Fi`, `Romance`]);

it(`Catalog genres list renders`, () => {
  const tree = renderer
    .create(
        <CatalogGenresList
          currentGenre={currentGenre}
          genres={genres}
          onChangeTab={noOperation}
          onGenreChange={noOperation}
          activeTab={`All genres`}
        />
    ).toJSON();

  expect(tree).toMatchSnapshot();
});
