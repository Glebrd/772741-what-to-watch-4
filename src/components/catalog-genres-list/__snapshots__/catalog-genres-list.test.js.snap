// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Catalog genres list renders 1`] = `
<ul
  className="catalog__genres-list"
>
  <li
    className="catalog__genres-item "
  >
    <a
      className="catalog__genres-link"
      href="#"
      onClick={[Function]}
    >
      Drams
    </a>
  </li>
  <li
    className="catalog__genres-item "
  >
    <a
      className="catalog__genres-link"
      href="#"
      onClick={[Function]}
    >
      Sci-Fi
    </a>
  </li>
  <li
    className="catalog__genres-item "
  >
    <a
      className="catalog__genres-link"
      href="#"
      onClick={[Function]}
    >
      Romance
    </a>
  </li>
</ul>
`;
